src/main/java/io/micronaut/build/InternalStateCheckingService.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2003-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.build;
+
+import org.gradle.api.logging.Logging;
+import org.gradle.api.provider.Property;
+import org.gradle.api.services.BuildService;
+import org.gradle.api.services.BuildServiceParameters;
+import org.slf4j.Logger;
+
+public abstract class InternalStateCheckingService implements BuildService<InternalStateCheckingService.Params>, AutoCloseable {
+    private static final Logger LOGGER = Logging.getLogger(InternalStateCheckingService.class);
+    public static final String NAME = "internalStateCheckingService";
+
+    interface Params extends BuildServiceParameters {
+        Property<Boolean> getRegisteredByProjectPlugin();
+    }
+
+    @Override
+    public void close() {
+        Boolean state = getParameters().getRegisteredByProjectPlugin().get();
+        if (Boolean.TRUE.equals(state)) {
+            LOGGER.warn("WARNING!\n" +
```

Comment:
Ouch, can we not use Java 17 already and use text blocks?

Reply:
For now I refrained on upgrading the plugin itself to use Java 17, so that it's easier to test my changes on projects which are not migrated yet (that is to say, all Micronaut projects :D)